### PR TITLE
[ABANDONED] EP/MIN-256. Supply and borrow APY.

### PR DESCRIPTION
**Pull request by Apollo50**
**Description:**

Add RPC to get total supply and borrow apy.
Refactor controller RPC functions.
Fix fn get_liquidity_pool_borrow_and_supply_rates().
Add tests.

**Changes in storage or API:**

Describe storage structure changes or API changes. Make sure to update the [API docs](https://minterestfinance.atlassian.net/wiki/spaces/MINTEREST/pages/134807557/Minterest+API).

**Checklist**

- [x] Code is covered with tests
- [ ] Code is covered with benchmarks
- [ ] Fresh benchmark results included

### CODE REVIEW

File: pallets/controller/src/lib.rs [hunk 1]
```diff
@@ -771,6 +775,89 @@ impl<T: Config> Pallet<T> {
 
 		Ok(interest_factor)
 	}
+
+	/// Calculate actual supply balance for user per asset based on fresh exchange rate.
+	///
+	/// - `who`: the AccountId whose balance should be calculated.
+	/// - `underlying_asset_id`: ID of the currency, the balance of supplying of which we calculate.
+	fn get_user_supply_per_asset(who: &T::AccountId, underlying_asset_id: CurrencyId) -> BalanceResult {
```
Comment (lut1k) [inline pallets/controller/src/lib.rs:783]: 1.Maybe it makes sense to place the methods `get_user_supply_per_asset()` and `get_user_borrow_per_asset()`  next to each other?
2. This comment is not entirely about your request. But could you fix the names of the parameters in the comments in the method `calculate_interest_params()` on lines 671-678? A camelCase is used there, but you need to use a snake_case.

**Comment by lut1k** [comment]
This comment is not entirely about your PR. But could you fix the names of the parameters in the comments in the method `calculate_interest_params()` on lines 671-678? A camelCase is used there, but you need to use a snake_case.

File: pallets/controller/src/lib.rs [hunk 2]
```diff
@@ -457,62 +458,15 @@ impl<T: Config> Pallet<T> {
...
-						}
-						Ok((
-							current_supply_in_usd + supply_in_usd,
-							current_borrowed_in_usd + borrowed_in_usd,
-						))
-					},
-				)?;
+		let total_supply_and_borrow_in_usd = Self::get_vec_of_supply_and_borrow_balance(who)?;
+
+		let (total_supply_balance, total_borrowed_balance) = total_supply_and_borrow_in_usd.into_iter().fold(
```
Comment (lut1k) [inline pallets/controller/src/lib.rs:463]: I think we need to use the method `try_fold()` and `checked_add()`.

File: pallets/controller/src/lib.rs [hunk 3]
```diff
@@ -561,19 +502,82 @@ impl<T: Config> Pallet<T> {
...
+		// Get vector of balances for each asset(in USD).
+		let total_supply_and_borrow_per_asset = Self::get_vec_of_supply_and_borrow_balance(&who)?;
+
+		// Calculate total supply and borrow balance.
+		let (total_supply, total_borrow) = total_supply_and_borrow_per_asset.iter().fold(
+			(Balance::zero(), Balance::zero()),
+			|(total_supply, total_borrow), &(_, supply_balance, borrow_balance)| -> (Balance, Balance) {
+				(total_supply + supply_balance, total_borrow + borrow_balance)
+			},
+		);
```
Comment (lut1k) [inline pallets/controller/src/lib.rs:535]:
```suggestion
		let (total_supply, total_borrow) = Self::get_total_supply_and_borrowed_usd_balance(&who)?;
```

File: pallets/controller/src/lib.rs [hunk 4]
```diff
@@ -771,6 +775,89 @@ impl<T: Config> Pallet<T> {
...
+							.map(|x| x.into_inner())
+							.ok_or(Error::<T>::BalanceOverflow)?;
+					}
+					if !borrow_balance.is_zero() {
+						borrow_balance = Rate::from_inner(borrow_balance)
+							.checked_mul(&oracle_price)
+							.map(|x| x.into_inner())
+							.ok_or(Error::<T>::BalanceOverflow)?;
+					}
+
```
[the remark this answers is not in the document]
Comment (Apollo50) [reply pallets/controller/src/lib.rs:846]: I don't like it too, but it allows do not multiply zero balance to oracle_price.

File: pallets/controller/src/lib.rs [hunk 5]
```diff
@@ -457,62 +458,15 @@ impl<T: Config> Pallet<T> {
 	pub fn get_total_supply_and_borrowed_usd_balance(
```
[the remark this answers is not in the document]
Comment (Apollo50) [reply pallets/controller/src/lib.rs:458]: Nice suggestion.
Fixed

File: pallets/controller/src/lib.rs [hunk 6]
```diff
@@ -771,6 +775,89 @@ impl<T: Config> Pallet<T> {
...
+	/// Create a vector that contains total supply and total borrowed balance in usd for each pool
+	/// based on freshest data.
+	///
+	/// - `who`: the AccountId whose APYs should be calculated.
+	/// Returns: Vec<(CurrencyId, supply_balance, borrow_balance)>
+	fn get_vec_of_supply_and_borrow_balance(
+		who: &T::AccountId,
+	) -> result::Result<Vec<(CurrencyId, Balance, Balance)>, DispatchError> {
+		CurrencyId::get_enabled_tokens_in_protocol(UnderlyingAsset)
+			.iter()
```
[the remark this answers is not in the document]
Comment (Apollo50) [reply pallets/controller/src/lib.rs:816]: fixed

File: pallets/controller/src/lib.rs (hunk 2 above)
Comment (Apollo50) [reply pallets/controller/src/lib.rs:463]: I also think so.

File: pallets/controller/src/lib.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (lut1k) [reply pallets/controller/src/lib.rs:855]: In my opinion, this optimization does not give us any improvement. And the code reads worse.
Comparison operations also consume CPU.

[the remark this answers is not in the document]
Comment (lut1k) [reply pallets/controller/src/lib.rs:855]: The options are:
- the user has collateral and a borrow -> we have 4 extra comparison operations. We are losing productivity.
- the user does not have a borrow and collateral -> we have two comparison operations, we do not read from the store. We've got the benefit of reading from the storage..
- the user has a borrow or collateral -> we have 2 extra comparison operations.

File: pallets/controller/src/lib.rs [hunk 7]
```diff
@@ -771,6 +784,89 @@ impl<T: Config> Pallet<T> {
...
+			.ok_or(Error::<T>::NumOverflow)?;
+		Ok(supply_balance)
+	}
+
+	/// Create a vector that contains total supply and total borrowed balance in usd for each pool
+	/// based on freshest data.
+	///
+	/// - `who`: the AccountId whose APYs should be calculated.
+	/// Returns: Vec<(CurrencyId, supply_balance, borrow_balance)>
+	fn get_vec_of_supply_and_borrow_balance(
```
Comment (lut1k) [inline pallets/controller/src/lib.rs:821]: I really don't like this function. We use this function twice in the protocol. And in both cases, we have an additional cycle for currencies.
In my opinion, this functionality can be implemented directly in these functions and get rid of the extra loop.

File: pallets/controller/src/lib.rs [hunk 8]
```diff
@@ -771,6 +784,89 @@ impl<T: Config> Pallet<T> {
...
+					vec.push((pool_id, supply_balance, borrow_balance));
+					Ok(vec)
+				},
+			)
+	}
+
+	/// Calculate the delta of blocks from the last interest accrued to the current.
+	///
+	/// - `underlying_asset_id`: ID of the currency, the blocks delta of which we calculate.
+	fn get_block_delta(underlying_asset_id: CurrencyId) -> result::Result<T::BlockNumber, DispatchError> {
```
Comment (lut1k) [inline pallets/controller/src/lib.rs:865]: 1. If we have implemented this function, then we can combine the functions `get_block_delta()` and `calculate_block_delta()` into one.
2. In the function `accrue_interest_rate()` we can use this new function `get_block_delta()`.

File: runtime/src/tests/rpc_tests.rs [hunk 9]
```diff
@@ -781,3 +781,81 @@ fn get_unclaimed_mnt_balance_should_work() {
...
+			);
+			// borrow_rate_per_year = 0,0000000072 × 5256000 = 3.78 %
+			// supply_rate_per_year = 0,000000005184 × 5256000 = 2.72 %
+			assert_eq!(
+				Controller::get_liquidity_pool_borrow_and_supply_rates(ETH),
+				Some((Rate::from_inner(7200000000), Rate::from_inner(5184000000)))
+			);
+
+			// Hypothetical year supply interest(for the pool):
+			// supplyInterest = userSupplyInUsd * supplyApyAsDecimal
```
Comment (lut1k) [inline runtime/src/tests/rpc_tests.rs:821]: snake_case ><

Comment (ales-by) [reply runtime/src/tests/rpc_tests.rs:821]: fixed

File: pallets/controller/src/tests.rs [hunk 10]
```diff
@@ -334,15 +334,204 @@ fn get_liquidity_pool_borrow_and_supply_rates_above_kink() {
 		.build()
 		.execute_with(|| {
 			// utilization_rate = 500 / (100 - 0 + 500) = 0.83 > kink = 0.8
-			// borrow_rate = 0.83 * 0.8 * 0.000_000_207  + (0.8 * 0.000_000_009) + 0 = 0.0000001452
-			// supply_rate = 0.83 * 0.0000001452 * (1 - 0.1) = 0.00000000455625
+			// borrow_rate = 0,83333336 * 0.8 * 0.000_000_207  + (0.8 * 0.000_000_009) + 0 = 0.0000001452
+			// supply_rate = 0,83333336 * 0.0000001452 * (1 - 0.1) = 0,0000001089
```
[the remark this answers is not in the document]
Comment (ales-by) [reply pallets/controller/src/tests.rs:338]: fixed

**Pull request closed without merging**